README.md
# BrainLove

Ruby implementation of [brainfuck](http://esolangs.org/wiki/Brainfuck).

## Installation

Install it as:

```
$ gem install brain_love
```

## Usage

### Standalone

```
# Hello World
$ brain_love -e '>+++++++++[<++++++++>-]<.>+++++++[<++++>-]<+.+++++++..+++.>>>++++++++[<++++>-]<.>>>++++++++++[<+++++++++>-]<---.<<<<.+++.------.--------.>>+.'
```

Or from file:

```
$ brain_love some_file.bf
```

### In your code

```ruby
require 'brain_love'

# Hello World
BrainLove.run_string('>+++++++++[<++++++++>-]<.>+++++++[<++++>-]<+.+++++++..+++.>>>++++++++[<++++>-]<.>>>++++++++++[<+++++++++>-]<---.<<<<.+++.------.--------.>>+.')
```

## Portability

 * Data space contains 30 000 cells
 * Data cell is unsigned byte (wraps around)
 * EOF sets data cell to 0
 * Nested loops
 * All unknown characters are considered as comments

## Current limitations

 * jump offsets are 1 byte

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## License

Copyright (c) 2012 Мар'ян Крекотень (Marjan Krekoten'). Distributed under the MIT License. See LICENSE for further details.

CHANGELOG.md
# Changelog

## v0.0.3

 * Parser and transformer were rewritten. All unknown characters are ignored and considered as comments.

## v0.0.2

 * Data cells now are unsigned byte and wrap around on decrement and increment

## v0.0.1

 * Beginning of fun


spec/spec_helper.rb
require 'brain_love'

def fixture_path(name)
  File.expand_path("../fixtures/#{name}", __FILE__)
end

RSpec::Matchers.define :parse do |string_for_parsing|
  match do |parser|
    begin
      parser.parse(string_for_parsing)
      true
    rescue Parslet::ParseFailed, NotImplementedError => e
      @error = parser.error_tree
      false
    end
  end

  failure_message_for_should do |parser|
    "expected #{parser} to parse #{string_for_parsing}\n#{@error}"
  end

  failure_message_for_should_not do |parser|
    "didn't expect #{parser} to parse #{string_for_parsing}"
  end

  description do
    "parse #{string_for_parsing}"
  end
end

RSpec::Matchers.define :transform do |string_for_transforming|
  match do |transformer|
    @intermediate_representation = transformer.apply(@raw_ir = BrainLove::Parser.new.parse(string_for_transforming))
    @node == @intermediate_representation
  end

  chain :to do |node|
    @node = node
  end

  failure_message_for_should do |transformer|
    "expected #{string_for_transforming}\n" <<
    "to be transformed to\n#{@node.inspect}\n" <<
    "got\n#{@intermediate_representation.inspect}\nir:\n#{@raw_ir}"
  end

  description do
    "transform #{string_for_transforming} to AST node"
  end
end


spec/brain_love_spec.rb
require 'spec_helper'
require 'stringio'

describe BrainLove do
  describe '#run_string' do
    it 'executes code from string' do
      output = StringIO.new
      BrainLove.run_string(
        '>+++++++++[<++++++++>-]<.>+++++++[<++++>-]<+.+++++++..+++.>>>++++++++[<++++>-]<.>>>++++++++++[<+++++++++>-]<---.<<<<.+++.------.--------.>>+.',
        nil,
        output
      )
      output.string.should == 'Hello World!'
    end
  end

  describe '#run_file' do
    it 'executes code from file' do
      output = StringIO.new
      BrainLove.run_file(
        fixture_path('hello_world.bf'),
        nil,
        output
      )
      output.string.should == 'Hello World!'
    end
  end
end


spec/brain_love/vm_spec.rb
require 'spec_helper'
require 'stringio'

describe BrainLove::VM do
  let(:input) { double('input stream') }
  let(:output) { double('output stream') }
  let(:vm) { BrainLove::VM.new(code, input, output) }

  describe "initial state" do
    let(:code) { [].join }

    it "sets IP to 0" do
      vm.ip.should == 0
    end

    it "sets DP to 0" do
      vm.dp.should == 0
    end
  end

  describe "INC_DP" do
    let(:code) { [BrainLove::VM::INC_DP].map(&:chr).join }

    it "increments the pointer by one" do
      expect { vm.execute }.to change { vm.dp }.by(1)
    end
  end

  describe "DEC_DP" do
    let(:code) { [BrainLove::VM::INC_DP, BrainLove::VM::DEC_DP].map(&:chr).join }

    it "decrements the pointer by one" do
      expect { vm.execute }.to change { vm.dp }.by(0)
    end
  end

  describe "INC_BYTE" do
    let(:code) { [BrainLove::VM::INC_DP, BrainLove::VM::INC_BYTE].map(&:chr).join }

    it "increments the byte at the pointer by one" do
      expect { vm.execute }.to change { vm.data[1] }.by(1)
    end

    it "wraps around value" do
      code = Array.new(256, BrainLove::VM::INC_BYTE).map(&:chr).join
      vm = BrainLove::VM.new(code, input, output)
      expect { vm.execute }.to change { vm.data[0] }.by(1)
    end
  end

  describe "DEC_BYTE" do
    let(:code) { [BrainLove::VM::INC_DP, BrainLove::VM::INC_BYTE, BrainLove::VM::INC_BYTE, BrainLove::VM::DEC_BYTE].map(&:chr).join }

    it "decrements the byte at the pointer by one" do
      expect { vm.execute }.to change { vm.data[1] }.by(1)
    end

    it "wraps around value" do
      code = BrainLove::VM::DEC_BYTE.chr
      vm = BrainLove::VM.new(code, input, output)
      expect { vm.execute }.to change { vm.data[0] }.by(255)
    end
  end

  describe "PUTC" do
    let(:code) { ([BrainLove::VM::INC_BYTE] * 97 << BrainLove::VM::PUTC).map(&:chr).join }

    it "sends the byte at the pointer as char to output stream" do
      output.should_receive(:putc).with("a")
      vm.execute
    end
  end

  describe "GETC" do
    let(:code) { [BrainLove::VM::INC_DP, BrainLove::VM::GETC].map(&:chr).join }

    before do
      input.should_receive(:getc).and_return("a")
    end

    it "reads char from input stream" do
      vm.execute
    end

    it "stores char as a byte at the pointer" do
      vm.execute
      vm.data[1].should == 97
    end
  end

  describe "JMPFZ" do
    let(:code) { [BrainLove::VM::JMPFZ, 3, BrainLove::VM::INC_BYTE, BrainLove::VM::INC_DP, BrainLove::VM::GETC, BrainLove::VM::PUTC].map(&:chr).join }

    it "jumps forward if current byte is 0" do
      input.stub(:getc).and_return('t')
      output.should_receive(:putc).with('t')
      expect { vm.execute }.to change { vm.data[0] }.by(0)
    end
  end

  describe "JMPBNZ" do
    let(:code) { [BrainLove::VM::INC_BYTE, BrainLove::VM::DEC_BYTE, BrainLove::VM::JMPBNZ, 0, BrainLove::VM::INC_BYTE].map(&:chr).join }

    it "jumps bacward unless current byte is 0" do
      expect { vm.execute }.to change { vm.data[0] }.by(1)
    end
  end
end


lib/brain_love/parser.rb
require 'parslet'

module BrainLove
  class Parser < Parslet::Parser
    rule(:increment_pointer)  { str('>').as(:increment_pointer) }
    rule(:decrement_pointer)  { str('<').as(:decrement_pointer) }
    rule(:increment_byte)     { str('+').as(:increment_byte) }
    rule(:decrement_byte)     { str('-').as(:decrement_byte) }
    rule(:output_byte)        { str('.').as(:output_byte) }
    rule(:input_byte)         { str(',').as(:input_byte) }
    rule(:jump_forward)       { str('[') }
    rule(:jump_backward)      { str(']') }

    rule :command do
      increment_pointer |
      decrement_pointer |
      increment_byte    |
      decrement_byte    |
      output_byte       |
      input_byte
    end

    rule :comment do
      (command | jump_forward | jump_backward).absnt? >> any.as(:comment)
    end

    rule :_loop do
      (jump_forward >> root >> jump_backward).as(:loop)
    end

    rule :root do
      (comment | command | _loop).repeat.as(:statements)
    end
  end
end


lib/brain_love/compiler.rb
module BrainLove
  class Compiler

    attr_reader :bytecode, :ip

    def initialize
      @bytecode = ""
      @ip = 0
    end

    def inc_dp
      @bytecode << VM::INC_DP.chr
    end

    def dec_dp
      @bytecode << VM::DEC_DP.chr
    end

    def inc_byte
      @bytecode << VM::INC_BYTE.chr
    end

    def dec_byte
      @bytecode << VM::DEC_BYTE.chr
    end

    def putc
      @bytecode << VM::PUTC.chr
    end

    def getc
      @bytecode << VM::GETC.chr
    end

    def jmpfz(at, offset)
      @bytecode[at] = VM::JMPFZ.chr
      @bytecode[at + 1] = offset.chr
    end

    def jmpbnz(offset)
      @bytecode << VM::JMPBNZ.chr << offset.chr
    end

    def noop
      @bytecode << VM::NOOP
    end

    def current_ip
      @bytecode.bytesize
    end

    def visit_Statements(statements)
      statements.value.each { |st| st.accept(self) }
    end

    def visit_Loop(_loop)
      jmpf_ip = current_ip
      noop
      noop
      _loop.value.accept(self)
      jmpbnz(current_ip - jmpf_ip)
      jmpfz(jmpf_ip, current_ip)
    end

    def visit_DecrementPointer(_)
      dec_dp
    end

    def visit_IncrementPointer(_)
      inc_dp
    end

    def visit_IncrementByte(_)
      inc_byte
    end

    def visit_DecrementByte(_)
      dec_byte
    end

    def visit_OutputByte(_)
      putc
    end

    def visit_InputByte(_)
      getc
    end
  end
end


lib/brain_love/vm.rb
module BrainLove
  class VM

    NOOP      = 0
    INC_DP    = 1
    DEC_DP    = 2
    INC_BYTE  = 3
    DEC_BYTE  = 4
    PUTC      = 5
    GETC      = 6
    JMPFZ     = 7
    JMPBNZ    = 8

    attr_reader :ip, :dp, :code, :data

    def initialize(code, input, output)
      @code, @input, @output = code, input, output
      @ip = 0
      @dp = 0
      @data = [0] * 30_000
    end

    def code_dump
      index = 0, jmp = false
      @code.each_byte do |b|
        print "#{index} "
        index += 1
        case b
        when NOOP
          puts "NOOP"
        when INC_DP
          puts "INC_DP"
        when DEC_DP
          puts "DEC_DP"
        when INC_BYTE
          puts "INC_BYTE"
        when DEC_BYTE
          puts "DEC_BYTE"
        when PUTC
          puts "PUTC"
        when GETC
          puts "GETC"
        when JMPFZ
          puts "JMPFZ"
          jmp = true
        when JMPBNZ
          puts "JMPBNZ"
          jmp = true
        else
          if jmp
            jmp = false
            puts " offset #{b}"
          else
            puts "UNKNOWN BYTECODE"
          end
        end
      end
    end

    def execute
      while @ip < @code.bytesize do
        case @code[@ip].ord
        when NOOP
        when INC_DP
          @dp += 1
        when DEC_DP
          @dp -= 1
        when INC_BYTE
          @data[@dp] = 0 if @data[@dp] == 255
          @data[@dp] += 1
        when DEC_BYTE
          @data[@dp] = 256 if @data[@dp] == 0
          @data[@dp] -= 1
        when PUTC
          @output.putc(@data[@dp].chr)
        when GETC
          @data[@dp] = (@input.getc || 0).ord
        when JMPFZ
          @ip += @code[@ip + 1].ord - 1 if @data[@dp] == 0
        when JMPBNZ
          unless @data[@dp] == 0
            @ip -= @code[@ip + 1].ord
          else
            @ip += 1
          end
        end

        @ip += 1
      end
    end
  end
end
